fragment.go

```diff
@@ -741,11 +790,26 @@ func (f *fragment) importSetValue(columnID uint64, bitDepth uint, value uint64,
 	}
 
 	// Mark value as set.
-	p, err := f.pos(uint64(bitDepth), columnID)
-	if err != nil {
+	if p, err := f.pos(uint64(bsiExistsBit), columnID); err != nil {
 		return changed, errors.Wrap(err, "getting not-null pos")
+	} else if clear {
+		if c, err := f.storage.Remove(p); err != nil {
+			return changed, errors.Wrap(err, "removing not-null from storage")
+		} else if c {
+			changed = true
+		}
+	} else {
+		if c, err := f.storage.Add(p); err != nil {
+			return changed, errors.Wrap(err, "adding not-null to storage")
+		} else if c {
+			changed = true
+		}
 	}
-	if clear {
+
+	// Mark sign bit.
+	if p, err := f.pos(uint64(bsiSignBit), columnID); err != nil {
+		return changed, errors.Wrap(err, "getting not-null pos")
+	} else if value >= 0 || clear {
 		if c, err := f.storage.Remove(p); err != nil {
 			return changed, errors.Wrap(err, "removing not-null from storage")
```

Comment:
it looks like this message should be about the sign bit now, not the not-null bit?

Reply:
+1